### DISCUSSION TITLE
Contributing to Uno Platform

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#18432] [Q&A] [answered]
I'm trying to solve an issue, for that I need to add a helper static class to the Uno.UI.Helper namespace and some modifications to each Uno.UI.Runtime.Skia.xxx projects.

My `crosstargeting_override.props` 

```xml
    <UnoTargetFrameworkOverride>net8.0</UnoTargetFrameworkOverride>
    <AccelerateBuildsInVisualStudio>true</AccelerateBuildsInVisualStudio>
    <!--<OptimizeImplicitlyTriggeredBuild>true</OptimizeImplicitlyTriggeredBuild>-->
    <UnoNugetOverrideVersion>5.5.0-dev.82</UnoNugetOverrideVersion>
```

First I opened `Uno.UI-Reference-Only.slnf`, I added the class and I built the solution, next I used `Uno.UI.Skia.only.slnf` to modify each skia runtime project and again I built the solution.

According to the output my nuget cache has been overridden:
```
20>Uno.UI.Reference -> C:\Code\uno\src\Uno.UI\bin\Uno.UI.Reference\Debug\net8.0\Uno.UI.dll
20>OVERRIDING NUGET PACKAGE CACHE: C:\Users\mcnets\.nuget\packages\\uno.winui\5.5.0-dev.82\lib\net8.0 and C:\Users\mcnets\.nuget\packages\\uno.winui\5.5.0-dev.82\ref\net8.0
40>Uno.UI.Runtime.Skia.Wpf -> C:\Code\uno\src\Uno.UI.Runtime.Skia.Wpf\bin\Debug\net8.0\Uno.UI.Runtime.Skia.Wpf.dll
40>OVERRIDING NUGET PACKAGE CACHE: C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.wpf\5.5.0-dev.82\lib\net8.0 and C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.wpf\5.5.0-dev.82\ref\net8.0
37>OVERRIDING NUGET PACKAGE CACHE: C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.x11\5.5.0-dev.82\lib\net8.0 and C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.x11\5.5.0-dev.82\ref\net8.0
37>OVERRIDING NUGET PACKAGE CACHE: C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.x11\5.5.0-dev.82\lib\net8.0
39>OVERRIDING NUGET PACKAGE CACHE: C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.linux.framebuffer\5.5.0-dev.82\lib\net8.0 and C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.linux.framebuffer\5.5.0-dev.82\ref\net8.0
39>OVERRIDING NUGET PACKAGE CACHE: C:\Users\mcnets\.nuget\packages\\uno.winui.runtime.skia.linux.framebuffer\5.5.0-dev.82\lib\net8.0
```

Next I created a new Uno project, setting `"Uno.Sdk": "5.5.0-dev.82"` but it is not getting the proper libaries.

Having a look at my nuget cache:
```
    Directory: C:\Users\mcnets\.nuget\packages\uno.ui\5.5.0-dev.82\lib\net8.0

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          03/10/2024    11:33       11303936 Uno.UI.dll
-a---          03/10/2024    11:33          28456 Uno.UI.dll.config
-a---          03/10/2024    11:33        4826792 Uno.UI.pdb
```
The library has not been updated, today is 10/10/2024.

What am I doing wrong?

**Comment by mcNets** [comment]
Now I'm noticing that is using `uno.winui` instead of `uno.ui`

Where should I add my class to get a new `Uno.UI.dll`?

**Reply by jeromelaban** [reply]
Thanks for the question. Is your project using `Uno.SDK` or `Uno.UI`/`Uno.WinUI` directly? The sources from the uno repo are configured to target the `uno.sdk`/`uno.winui` packages, not the `uno.ui` package.

**Reply by mcNets** [reply]
@jeromelaban  My test project has been generated with the Uno template. Extension: 5.4.13.15
```
{
  // To update the version of Uno please update the version of the Uno.Sdk here. See https://aka.platform.uno/upgrade-uno-packages for more information.
  "msbuild-sdks": {
    "Uno.Sdk": "5.5.0-dev.82"
  },
  "sdk":{
    "allowPrerelease": false
  }
}
```
For the Uno project I forked your repo, modified `crosstargeting_override.props` and used the 2 filtered solutions I mentioned above.

[This](https://github.com/unoplatform/uno/compare/master...mcNets:uno:platform-runtime-helper) is my branch.

And [this](https://github.com/unoplatform/uno/issues/11545) is the issue I'm trying to solve.

**Reply by morning4coffe-dev** [reply, marked as answer]
@mcNets It looks like this issue is caused by a mismatch between the Uno.Sdk and Uno.WinUI versions. In your case, to override Sdk 5.5.0-dev.82, you will need to do:

```xml
<UnoNugetOverrideVersion>5.5.0-dev.573</UnoNugetOverrideVersion>
```

You can find the corresponding versions [here](https://www.nuget.org/packages/Uno.Sdk/5.5.0-dev.82#readme-body-tab:~:text=Default%20Version-,UnoVersion,-*).

**Reply by mcNets** [reply]
@jeromelaban @morning4coffe-dev  Oh, thank you. I thought both versions were the same.

**Reply by mcNets** [reply]
@morning4coffe-dev  I need to test all platforms, what other filtered solutions should I rebuild?

![image](https://github.com/user-attachments/assets/72e56128-a714-432d-bfc8-e6d0370c4497)

And what's the difference between `uno.winui` and `uno.ui` packages?

I suppose there is a uno.dll specific for each platform, but then why a `uno.ui` package?

**Reply by MartinZikmund** [reply]
Uno.WinUI is with Microsoft.UI.Xaml.XYZ namespaces, so WinUI-based, Uno.UI is WUX namespaces, so UWP based. Technically you shouldn't need to adjust anything for this PR.

To test all platforms, you will need to build the specific targets. So for instance for Android, you will need to keep the NuGet override and rebuild Uno.UI under `net8.0-android`, and so on.

However, if you want to have this API available on WinUI as well, then you will need to put it inside Uno.UI.Toolkit project (not to be confused with the Uno Toolkit library, which is not part of this repository), because the Uno.UI.csproj is only compiled against non-WinUI targets.

**Reply by mcNets** [reply]
@MartinZikmund sorry🙏 ,  just another question, there are five of them, should I add the same class to each Toolkit project?

![image](https://github.com/user-attachments/assets/4a7ba1dd-0b07-4176-bdb1-dd2bbe1f2691)

**Reply by mcNets** [reply]
I see, it is automatically added to the other projects.

**Reply by MartinZikmund** [reply]
Yes, technically it is a single folder, just each csproj targets a different target

**Reply by mcNets** [reply]
@MartinZikmund  it seems that `Uno.UI.Runtime.Skia.XXX` does not include `Uno.UI.Toolkit`. Do you think it should be added?

**Reply by MartinZikmund** [reply]
As I think about it more, I think the best approach would be to make the class you already have `static internal` class in Uno.UI,, that will have the property with setter and then just have a `public` version of that helper inside of `Uno.UI.Toolkit` - this way the toolkit version will be public and will be accessible to everyone + will be able to recognize WinUI as well, while the helper can still also be used internally withing Uno itself

**Reply by mcNets** [reply]
@MartinZikmund hummm, where can I put the `enum UnoRuntimePlatform` to make it visible for the windows project?

Uno.UI.Helpers are no visible for windows.

```csharp
namespace Uno.UI.Toolkit;

public static class RuntimeHelper
{
	public static UnoRuntimePlatform CurrentPlatform =>
#if !HAS_UNO
		UnoRuntimePlatform.Windows;
#else
		PlatformRuntimeHelper.Current;
#endif
}
```

It seems a vicious circle.

**Reply by MartinZikmund** [reply]
Indeed, it is not ideal, technically you will have two "copies" of the same enum, but it is the only approach that works in this case - you will return the "internal" enum value from Uno.UI property and then a public enum value from the Toolkit class. They can be even named the same, just have a different namespace and one of them will be public
